hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerMerkleTree.java

```diff
@@ -0,0 +1,180 @@
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+/**
+ * This class represents a Merkle tree that provides one checksum for all data within a container.
+ *
+ * As the leaves of the tree, a checksum for each chunk is computed by taking a checksum of all checksums within that
+ * chunk. Each chunk checksum in a block is further checksummed together to generate the block level checksum. Finally,
+ * The checksums of all blocks are checksummed together to create a container level checksum.
+ * Note that checksums are order dependent. Chunk checksums are sorted by their
+ * offset within a block, and block checksums are sorted by their block ID.
+ *
+ * This class can be used to construct a consistent and completely filled {@link ContainerProtos.ContainerMerkleTree}
+ * object. It allows building a container merkle tree from scratch by incrementally adding chunks.
+ * The final checksums at higher levels of the tree are not calculated until
+ * {@link ContainerMerkleTree#toProto} is called.
+ */
+public class ContainerMerkleTree {
+
+  private final SortedMap<Long, BlockMerkleTree> id2Block;
+
+  /**
+   * Constructs an empty Container merkle tree object.
+   */
+  public ContainerMerkleTree() {
+    id2Block = new TreeMap<>();
+  }
+
+  /**
+   * Adds chunks to a block in the tree. The block entry will be created if it is the first time adding chunks to it.
+   * If the block entry already exists, the chunks will be added to the existing chunks for that block.
+   *
+   * @param blockID The ID of the block that these chunks belong to.
+   * @param chunks A list of chunks to add to this block. The chunks will be sorted internally by their offset.
+   */
+  public void addChunks(long blockID, Collection<ChunkInfo> chunks) {
+    id2Block.computeIfAbsent(blockID, BlockMerkleTree::new).addChunks(chunks);
+  }
+
+  /**
+   * Uses chunk hashes to compute all remaining hashes in the tree, and returns it as a protobuf object. No checksum
+   * computation for the tree happens outside of this method.
+   *
+   * @return A complete protobuf object representation of this tree.
+   */
+  public ContainerProtos.ContainerMerkleTree toProto() {
+    // Compute checksums and return the result.
+    ContainerProtos.ContainerMerkleTree.Builder containerTreeBuilder = ContainerProtos.ContainerMerkleTree.newBuilder();
+    ChecksumByteBuffer checksumImpl = ChecksumByteBufferFactory.crc32Impl();
+    ByteBuffer containerChecksumBuffer = ByteBuffer.allocate(Long.BYTES * id2Block.size());
+
+    for (BlockMerkleTree blockTree: id2Block.values()) {
+      ContainerProtos.BlockMerkleTree blockTreeProto = blockTree.toProto();
+      containerTreeBuilder.addBlockMerkleTree(blockTreeProto);
+      // Add the block's checksum to the buffer that will be used to calculate the container checksum.
+      containerChecksumBuffer.putLong(blockTreeProto.getBlockChecksum());
+    }
+    containerChecksumBuffer.flip();
+    checksumImpl.update(containerChecksumBuffer);
```

Comment:
It should be possible to take the longs of the blocks and just call update with the bits shifted to calculate the checksum.

Reply:
Yeah I can update this to use Java's CRC32 which has an `update(int)` method instead of our own `ByteBuffer` based implementation.

Reply:
After some offline discussion we are going to handle any optimizations in checksum computation in HDDS-11077